lib/lte_uart_dfu/lte_uart_dfu.c

```diff
@@ -0,0 +1,340 @@
+			lte_uart_dfu_send_error(
+				_DFU_UART_target_RemoteFnc_type_write, err);
+			break;
+		}
+
+		err = dfu_target_mcuboot_write(context.to_write,
+					       context.to_write_len);
+		progress += context.to_write_len;
+		LOG_INF("Progress: %d %%", progress * 100 / context.file_size);
+		lte_uart_dfu_send_error(_DFU_UART_target_RemoteFnc_type_write,
+					err);
+		break;
+	case _DFU_UART_target_RemoteFnc_type_offset:
+		err = dfu_target_mcuboot_offset_get(&offs_get);
+		lte_uart_dfu_send_offset(err, offs_get);
+		break;
+	case _DFU_UART_target_RemoteFnc_type_done:
+		err = lte_uart_dfu_parse_done(fnc);
+		if (err) {
+			lte_uart_dfu_send_error(
+				_DFU_UART_target_RemoteFnc_type_done, err);
+			break;
+		}
+
+		err = dfu_target_mcuboot_done(context.download_success);
+		if (err) {
+			lte_uart_dfu_send_error(
+				_DFU_UART_target_RemoteFnc_type_done, err);
+			break;
+		}
+
+		err = dfu_target_mcuboot_schedule_update(
+			context.image_num);
+		lte_uart_dfu_send_error(_DFU_UART_target_RemoteFnc_type_done,
+					err);
+		break;
+	default:
+		break;
+	}
+}
+
+static void process_input_data(void)
+{
+	int err;
+	uint8_t *cursor;
+	size_t cursor_len;
+	size_t dec_len = 0;
+	struct DFU_UART_target_RemoteFnc fnc = {0};
+
+	while (true) {
+		cursor_len = ring_buf_get_claim(&context.rx_rb, &cursor, UINT32_MAX);
+		if (cursor_len == 0) {
+			/* Let's reset it to prevent wrapping */
+			ring_buf_reset(&context.rx_rb);
+			return;
+		}
+
+		/* We try to put the whole buffer in the decode function.
+		 * If there is not enough data, the decode will throw an error.
+		 * If there is too much data, no error will be thrown and
+		 * the correct amount of decoded data will be given by the
+		 * decode function
+		 */
+		err = cbor_decode_DFU_UART_target_RemoteFnc(
+			cursor,
+			cursor_len,
+			&fnc,
+			&dec_len);
+		if (err) {
+			ring_buf_get_finish(&context.rx_rb, 0);
+			return;
+		}
+
+		execute_input_fnc(&fnc);
+
+		/* We need to remove the amount of consumed data */
+		err = ring_buf_get_finish(&context.rx_rb, dec_len);
+		if (err) {
+			LOG_ERR("ring_buf_get_finish returned %d", err);
+		}
+	}
+}
+
+noreturn static void uart_input_handler(void *p1, void *p2, void *p3)
+{
+	ARG_UNUSED(p1);
+	ARG_UNUSED(p2);
+	ARG_UNUSED(p3);
+
+	int err;
+
+	while (1) {
+		k_sem_take(&context.lte_uart_sem, K_FOREVER);
+		process_input_data();
+	}
+}
+
+int lte_uart_dfu_start(const struct device *uart)
+{
+	if (!uart) {
+		LOG_ERR("UART in not ready");
+		return -EINVAL;
+	}
+
+	context.uart_dev = uart;
+	context.out_buf_cursor = 0;
+
+	k_sem_init(&context.lte_uart_sem, 0, 1);
+	ring_buf_init(&context.rx_rb, ARRAY_SIZE(in_buf), in_buf);
+
+	tid = k_thread_create(&uart_handler_thread, uart_handler_stack,
+			K_THREAD_STACK_SIZEOF(uart_handler_stack),
+			uart_input_handler, NULL, NULL, NULL,
+			CONFIG_LTE_UART_DFU_HANDLER_PRIORITY, 0, K_NO_WAIT);
+	k_thread_name_set(&uart_handler_thread, "dfu_uart");
+
+	uart_irq_callback_set(uart, uart_cb);
```

Comment:
You could use `uart_irq_callback_user_data_set` here and pass your `context` here. Then in `uart_cb` you would get the context through the `p` pointer.

Reply:
Are there any gains of doing it?